File: algorithms/dynamic-programming/best-time-to-buy-and-sell-stock.py
class Solution:
  def maxProfit(self, prices: List[int]) -> int:
    length = len(prices)
    if length <= 1:
      return 0

    left_view = [0] * length
    right_view = [0] * length

    left_view[0] = prices[0]
    for i in range(1, length):
      left_view[i] = min(left_view[i-1], prices[i])

    right_view[length - 1] = prices[length - 1]
    for i in reversed(range(length - 1)):
      right_view[i] = max(right_view[i+1], prices[i])

    profit = 0
    for i in range(length):
      if right_view[i] - left_view[i] > profit:
        profit = right_view[i] - left_view[i]

    return profit

File: top-questions/Amazon-Top-Questions/dynamic-programming/best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
  let minprice = prices[0];
  let maxprofit = 0;

  for (let i = 0; i < prices.length; i++) {
    if (minprice > prices[i]) {
      minprice = prices[i];
    } else if (maxprofit < prices[i] - minprice) {
      maxprofit = prices[i] - minprice;
    }
  }

  return maxprofit;
};
